Add Vertex AI Full Version support

InterestingDarknessII opened the pull request:
<!-- Put X in the box below to confirm -->

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

AotianNeo committed:
Implement Vertex AI authentication modes and configuration in UI

- Updated index.html to include options for Vertex AI authentication modes (Express and Full).
- Enhanced openai.js to manage Vertex AI settings, including project ID, region, and service account JSON.
- Added validation and handling for service account JSON in the backend.
- Modified API request handling in google.js to support both authentication modes for Vertex AI.
- Updated secrets.js to include a key for storing Vertex AI service account JSON.
- Improved error handling and user feedback for authentication issues.

AotianNeo committed:
fix ESLint

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -471,6 +479,10 @@ const oai_settings = {
+    vertexai_auth_mode: 'express',
+    vertexai_project_id: '',
+    vertexai_region: 'us-central1',
+    vertexai_service_account_json: '',
```
~~Can we consider this to be a secret? Exposing a private key to frontend doesn't seem a good idea to me.~~

Nevermind, read further. Though, it doesn't belong to oai_settings anyway.

InterestingDarknessII replied:
https://github.com/SillyTavern/SillyTavern/pull/4046/commits/a9d4988fc000c0c7cb37d337b056d9b29225cfd4
Removed `vertexai_service_account_json`. Should we also store `project_id` in secrets?

Cohee1207 replied:
This also looks like a secret / sensitive info to me, so yeah.

Also, if you want to simplify the UI further, move the region to config.yaml (in gemini section, next to API version).

InterestingDarknessII replied:
I have submitted the modifications. https://github.com/SillyTavern/SillyTavern/pull/4046/commits/1e2bec1751ef06beb4b769dd290b7496b1df127a
I chose to keep the `region` textbox because the free `gemini-2.5-pro-exp-03-25` requires the `global` region, but other models respond faster when using other regions. I decided to give users the choice here.

Cohee1207 replied:
Is the list of regions known and static? Then how about making it a dropdown property instead of a text box?

InterestingDarknessII replied:
Different regions support different models; only `us-central1` supports all models.
https://cloud.google.com/vertex-ai/generative-ai/docs/learn/locations

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -387,6 +391,10 @@ const default_settings = {
     claude_use_sysprompt: false,
     use_makersuite_sysprompt: true,
+    vertexai_auth_mode: 'express',
+    vertexai_project_id: '',
```
I think those need to be added to sensitiveFields array, so a user would see a warning exporting/importing those in presets.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5480,6 +5527,160 @@ function runProxyCallback(_, value) {
+        $('#vertexai_full_config').hide();
+        $('#vertexai_express_models').show();
+        // Hide all full version model groups
+        $('#vertexai_full_gemini_25').hide();
```
I think we can handle this more gracefully by adding a data-mode attribute to optgroups instead. This will greatly simplify the code.
`<optgroup label="Express models" data-mode="express"></optgroup>`
`<optgroup label="Full models" data-mode="full"></optgroup>`

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -5480,6 +5527,160 @@ function runProxyCallback(_, value) {
+    // If no explicit message provided, check if we have a saved service account
+    if (!message && secret_state[SECRET_KEYS.VERTEXAI_SERVICE_ACCOUNT]) {
+        isValid = true;
+        message = 'Service Account JSON is saved and ready to use';
```
Consider making this localizable by using `t` template function.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -60,6 +61,124 @@ const API_DEEPSEEK = 'https://api.deepseek.com/beta';
+ * @param {object} serviceAccount Service account JSON object
+ * @returns {Promise<string>} JWT token
+ */
+async function generateJWTToken(serviceAccount) {
```
Please put those functions into `google.js` file since they are highly specific to google.

Cohee1207 replied:
Ah, they are already there. Then modify those to use export and reuse them here.

AotianNeo committed:
Refactor Google authentication functions for Vertex AI

- Exported `getVertexAIAuth`, `generateJWTToken`, and `getAccessToken` functions from google.js for better modularity.
- Removed redundant JWT token generation and access token retrieval logic from chat-completions.js, utilizing the newly exported functions instead.
- Improved code organization and maintainability by centralizing authentication logic.

AotianNeo committed:
Updated toastr messages in openai.js to use translation function for better internationalization support.

AotianNeo committed:
Enhance Vertex AI model selection by adding data-mode attributes

- Updated index.html to include data-mode attributes for model optgroups, distinguishing between express and full modes.
- Modified openai.js to show/hide model groups based on the selected authentication mode, improving user experience and functionality.

AotianNeo committed:
Add vertex AI configuration fields to sensitiveFields in openai.js

- Included 'vertexai_project_id', 'vertexai_region', and 'vertexai_service_account_json' to the sensitiveFields array for enhanced configuration management.

AotianNeo committed:
Remove vertexai_service_account_json from oai_settings. And google.js.

AotianNeo committed:
Removed direct references to 'vertexai_project_id' from openai.js and related files, ensuring it is now managed through backend secrets for enhanced security.

Cohee1207 reviewed (approved):
It's working. Though the process of setting it up is quite convoluted and probably requires a documentation page as it took me some time to figure out how to create a service account and that you don't actually need gcloud console app for it, and Google's own docs are horrendously bad at explaining it.

Also, the list of models likely requires cleaning up. 50% of models don't work for me even with the central-us region, mostly older ones from 1.5 and most of exp. snapshots. Can you please help me cleaning the list up?

AotianNeo committed:
Update Gemini model options in Vertex AI section

- Removed outdated model options from the Gemini 2.0 optgroup in index.html.
- Added 'gemini-2.0-flash-preview-image-generation' to the image generation models in chat-completions.js for improved functionality.

Cohee1207 commented on public/index.html:
```diff
@@ -3225,32 +3225,113 @@ <h4 data-i18n="Google Model">Google Model</h4>
+
+                                <!-- Project ID -->
+                                <div class="flex-container">
+                                    <label for="vertexai_project_id" data-i18n="Project ID">Project ID:</label>
```
Quick question, do we really need a separate control for Project ID? It's a required field of the auth JSON, and I'm sure you can NOT use a JSON from other project ID. Also LiteLLM doesn't require Project ID separately, only a JSON.

https://docs.litellm.ai/docs/providers/vertex

Cohee1207 replied:
I'd prefer going through the service account JSON method as it doesn't require to install any apps. 
If you still want to use the console app, we could add a disclaimer to add project_id field manually.

Cohee1207 commented on public/index.html:
```diff
@@ -3225,32 +3225,113 @@ <h4 data-i18n="Google Model">Google Model</h4>
+                                <!-- Service Account JSON Content -->
+                                <div class="flex-container flexFlowColumn">
+                                    <label for="vertexai_service_account_json" data-i18n="Service Account JSON Content">Service Account JSON Content:</label>
+                                    <textarea id="vertexai_service_account_json" class="text_pole textarea_compact" rows="8" placeholder='Paste your Service Account JSON content here, e.g.:
```
Let's probably make it a two or four rows textarea instead of 8. It appears too tall, and you're usually not entering it manually, just pasting a downloaded JSON from Google. So no point in making an elaborate placeholder

Cohee1207 commented on public/index.html:
```diff
@@ -3225,32 +3225,113 @@ <h4 data-i18n="Google Model">Google Model</h4>
+                                <!-- Region -->
+                                <div class="flex-container">
+                                    <label for="vertexai_region" data-i18n="Region">Region:</label>
+                                    <input id="vertexai_region" name="vertexai_region" class="text_pole flex1" value="us-central1" type="text" autocomplete="off" placeholder="e.g., us-central1, europe-west1, asia-northeast1">
```
I think we need to also make it obvious that "global" is an accepted value.

Also a link to docs with a list of models per region would be helpful.

Cohee1207 commented on public/index.html:
```diff
@@ -3225,32 +3225,113 @@ <h4 data-i18n="Google Model">Google Model</h4>
-                                    <optgroup label="Gemini 2.0">
+
+                                    <!-- Full Version Models -->
+                                    <optgroup id="vertexai_full_gemini_25" label="Gemini 2.5" data-mode="full" style="display: none;">
```
What would be expected effort to verify the functionality of models other than Gemini which are also hosted on Vertex? I see they also provide Anthropic, Jamba, Mistral, Llama, etc...

Cohee1207 replied:
Ok, leave it out for now. Can be figured out later.

AotianNeo committed:
Reduced the number of rows in the textarea for Service Account JSON input in index.html

AotianNeo committed:
Updated the placeholder text for the region input field in index.html to include 'global' as an option.
- Added a help link to provide users with information on available regions and models, improving user guidance.

AotianNeo committed:
Derive Vertex AI Project ID from Service Account JSON

This commit refactors the Vertex AI integration to automatically derive the
Project ID from the provided Service Account JSON. This simplifies the
configuration process for users in "Full" (service account) authentication
mode by removing the need to specify the Project ID separately.

Cohee1207 committed:
Fix model options hiding on Chrome

Cohee1207 committed:
Implement auto-connect for VertexAI full

Cohee1207 committed:
Remove unnecessary blank lines in chat-completions.js

Cohee1207 committed:
Fix alignment of Region controls

Cohee1207 committed:
Move account status block higher

Cohee1207 committed:
Change no neutral warning text to align with other inputs
